subcommand/create-federation-secret/command_test.go

```diff
@@ -0,0 +1,1118 @@
+		require.Equal(t, expCfg, string(secret.Data["serverConfigJSON"]))
+	}
+
+	// Now re-run the command.
+	{
+		// Update the mesh gateway IP.
+		newMeshGWIP := "127.0.0.1"
+		err = client.Agent().ServiceRegister(&api.AgentServiceRegistration{
+			Name: "mesh-gateway",
+			TaggedAddresses: map[string]api.ServiceAddress{
+				"wan": {
+					Address: newMeshGWIP,
+					Port:    meshGWPort,
+				},
+			},
+		})
+		require.NoError(t, err)
+
+		ui := cli.NewMockUi()
+		cmd := Command{
+			UI:        ui,
+			k8sClient: k8s,
+		}
+		exitCode := cmd.Run([]string{
+			"-resource-prefix", resourcePrefix,
+			"-k8s-namespace", k8sNS,
+			"-mesh-gateway-service-name=mesh-gateway",
+			"-ca-file", caFile,
+			"-server-ca-cert-file", caFile,
+			"-server-ca-key-file", keyFile,
+			"-http-addr", fmt.Sprintf("https://%s", a.HTTPSAddr),
+		})
+		require.Equal(t, 0, exitCode, ui.ErrorWriter.String())
+
+		// Check the secret is as expected.
+		secret, err := k8s.CoreV1().Secrets(k8sNS).Get(resourcePrefix+"-federation", metav1.GetOptions{})
+		require.NoError(t, err)
+
+		// Test server config. The mesh gateway IP should be updated.
+		require.Contains(t, secret.Data, "serverConfigJSON")
+		expCfg := fmt.Sprintf(`{"primary_datacenter":"dc1","primary_gateways":["%s:%d"]}`, newMeshGWIP, meshGWPort)
+		require.Equal(t, expCfg, string(secret.Data["serverConfigJSON"]))
+	}
+}
+
+// Test that if the Consul client isn't up yet we will retry until it is.
+func TestRun_ConsulClientDelay(t *testing.T) {
+	t.Parallel()
+
+	// We need to reserve all 6 ports to avoid potential
+	// port collisions with other tests.
+	randomPorts := freeport.MustTake(6)
+	caFile, certFile, keyFile, cleanup := common.GenerateServerCerts(t)
+	defer cleanup()
+
+	// Create fake k8s.
+	k8s := fake.NewSimpleClientset()
+
+	// Set up Consul server with TLS. Start after a 500ms delay.
+	var consulCleanup func() error
+	go func() {
+		time.Sleep(500 * time.Millisecond)
+		a, err := testutil.NewTestServerConfigT(t, func(cfg *testutil.TestServerConfig) {
+			cfg.CAFile = caFile
+			cfg.CertFile = certFile
+			cfg.KeyFile = keyFile
+			cfg.Ports = &testutil.TestPortConfig{
+				DNS:     randomPorts[0],
+				HTTP:    randomPorts[1],
+				HTTPS:   randomPorts[2],
+				SerfLan: randomPorts[3],
+				SerfWan: randomPorts[4],
+				Server:  randomPorts[5],
+			}
+		})
+		require.NoError(t, err)
+		consulCleanup = a.Stop
+
+		// Construct Consul client.
+		client, err := api.NewClient(&api.Config{
+			Address: a.HTTPAddr,
+		})
+		require.NoError(t, err)
+
+		// Create mesh gateway.
+		meshGWIP := "192.168.0.1"
+		meshGWPort := 443
+		err = client.Agent().ServiceRegister(&api.AgentServiceRegistration{
+			Name: "mesh-gateway",
+			TaggedAddresses: map[string]api.ServiceAddress{
+				"wan": {
+					Address: meshGWIP,
+					Port:    meshGWPort,
+				},
+			},
+		})
+		require.NoError(t, err)
+	}()
+	defer func() {
+		if consulCleanup != nil {
```

Comment:
If `consulCleanup` is nil here, then it could be because the consul test agent still hasn't finished starting up. At which point, we could be leaving it run in a goroutine forever. I think I ended up adding waitgroups to the `get-consul-client-ca` [command tests](https://github.com/hashicorp/consul-k8s/blob/0cec4d05429fa88ed14ea565fa5dd038de26cd3f/subcommand/get-consul-client-ca/command_test.go#L145-L175), but maybe there's a better option.

Reply:
Nice, I think that's a good solution.